### PR TITLE
[next] Support recommending similar E2E scenarios based on the recent multiple execution commands

### PR DESCRIPTION
**Pull request by ReaNAiveD**
Description:

- Add command arguments `--scenario` and `--command` to recommending scenarios or commands only. 
- Display commands and scenarios seperately when recommendation contains both commands and scenarios. 
- Display the commands the user has executed in scenario in secondary style. 

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
az next

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repository and upgrade the version in the pull request but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
az next improvement

File: src/next/azext_next/__init__.py [hunk 1]
```diff
@@ -25,8 +26,7 @@ def __init__(self, cli_ctx=None):
         from azure.cli.core.commands import CliCommandType
         next_custom = CliCommandType(
             operations_tmpl='azext_next.custom#{}')
-        super(NextCommandsLoader, self).__init__(cli_ctx=cli_ctx,
-                                                 custom_command_type=next_custom)
+        super().__init__(cli_ctx=cli_ctx, custom_command_type=next_custom)
```
Comment (zhoxing-ms) [inline src/next/azext_next/__init__.py:29]: Why do we need to modify here?

File: src/next/azext_next/_params.py [hunk 2]
```diff
@@ -8,4 +8,6 @@
 
 
 def load_arguments(self, _):  # pylint: disable=unused-argument
-    pass
+    with self.argument_context('next') as c:
+        c.argument('scenario_only', options_list=['--scenario', '-s'], action='store_true', help='recommend e2e scenarios only')
+        c.argument('command_only', options_list=['--command', '-c'], action='store_true', help='recommend commands only')
```
Comment (zhoxing-ms) [inline src/next/azext_next/_params.py:13]:
```suggestion
        c.argument('scenario_only', options_list=['--scenario', '-s'], action='store_true', help='Specify this parameter will only recommend scenario')
        c.argument('command_only', options_list=['--command', '-c'], action='store_true', help='Specify this parameter will only recommend commands')
```

File: src/next/azext_next/_help.py [hunk 3]
```diff
@@ -16,19 +16,22 @@
     [1] az config set next.execute_in_prompt=True/False
         Turn on/off the step of executing recommended commands in interactive mode. Turn on by default.
 
-    [2] az config set next.filter_type=True/False
-        Turn on/off the step of filtering recommendation type. Turn off by default.
+    [2] az config set next.filter_type=mix/scenario/command
+        Set default filtering recommendation type. Mix is the default.
```
Comment (zhoxing-ms) [inline src/next/azext_next/_help.py:20]:
```suggestion
    [2] az config set next.recommended_type=all/scenario/command
        Set default recommendation type. All is the default.
```

File: src/next/azext_next/custom.py [hunk 4]
```diff
@@ -2,60 +2,103 @@
...
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
-import re
 import json
+import re
 
+from azure.cli.core import telemetry
 from azure.cli.core.azclierror import RecommendationError
+from azure.cli.core.style import Style, print_styled_text
+from colorama import Fore, init
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:11]: Could we drop the usage of `colorama` as `Scenario Guide` does?

File: src/next/azext_next/custom.py [hunk 5]
```diff
@@ -2,60 +2,103 @@
...
-    if cmd.cli_ctx.config.getboolean('next', 'filter_type', fallback=False):
-        request_type = _get_filter_option()
+    if scenario_only:
+        request_type = RecommendType.Scenario.value
+    elif command_only:
+        request_type = RecommendType.Command.value
     else:
-        request_type = RecommendType.All.value
+        # Fallback to the configured filter_type if not command_only and scenario_only
+        request_type = RecommendType.get(cmd.cli_ctx.config.get('next', 'filter_type', fallback='mix')).value
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:30]:
```suggestion
        request_type = RecommendType.get(cmd.cli_ctx.config.get('next', 'recommended_type', fallback='all')).value
```

File: src/next/azext_next/custom.py [hunk 6]
```diff
@@ -2,60 +2,103 @@
...
-    print()
+    # divide recommendation items into two sets
+    command_recommendations = [item for item in recommends if item['type'] != RecommendType.Scenario]
+    scenario_recommendations = [item for item in recommends if item['type'] == RecommendType.Scenario]
+    if len(command_recommendations) == 0:
+        multi_type_recommendation = False
+    elif len(scenario_recommendations) == 0:
+        multi_type_recommendation = False
+    else:
+        multi_type_recommendation = True
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:62]:
```suggestion
        multi_type_recommendation = command_recommendations and scenario_recommendations
```

File: src/next/azext_next/__init__.py (hunk 1 above)
Comment (ReaNAiveD) [reply src/next/azext_next/__init__.py:29]: > Why do we need to modify here?

To fix a suggestion reported by `azdev style next`.

File: src/next/azext_next/custom.py [hunk 7]
```diff
@@ -307,15 +363,20 @@ def _execute_recommend_commands(cmd, rec):
 
 
 def _execute_recommend_scenarios(cmd, rec):
-    for nx_cmd in rec["nextCommandSet"]:
+    exec_idx = rec.get("executeIndex")
+    for idx, nx_cmd in enumerate(rec["nextCommandSet"]):
+        cmd_active = exec_idx is None or idx in exec_idx
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:368]: > exec_idx is None

We have filtered on the server side, why does this happen?

Comment (ReaNAiveD) [reply src/next/azext_next/custom.py:368]: > > exec_idx is None
> 
> We have filtered on the server side, why does this happen?

The scenario recommendations by knowledge base are based on the first command of scenario. In that case, there is no field `executeIndex`.

File: src/next/azext_next/custom.py [hunk 8]
```diff
@@ -368,13 +429,16 @@ def _show_details_for_e2e_scenario(cmd, rec):
...
-    for nx_cmd in nx_cmd_set:
-        idx += 1
+    exec_idx = rec.get("executeIndex")
+    for idx, nx_cmd in enumerate(nx_cmd_set):
         command_item = "az " + nx_cmd['command']
         if 'arguments' in nx_cmd and cmd.cli_ctx.config.getboolean('next', 'show_arguments', fallback=False):
-            command_item = "{} {}".format(command_item, ' '.join(nx_cmd['arguments']))
-        print(command_item)
+            command_item = f"{command_item} {' '.join(nx_cmd['arguments'])}"
+        cmd_active = exec_idx is None or idx in exec_idx
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:437]: > exec_idx is None

Same as above

File: src/next/azext_next/utils.py [hunk 9]
```diff
@@ -177,15 +240,16 @@ def get_last_exception(cmd, latest_command):
...
-def get_title_case(str):
-    if not str:
-        return str
-    str = str.strip()
-    return str[0].upper() + str[1:]
+def get_title_case(s):
+    if not s:
+        return s
+    s = s.strip()
+    return s[0].upper() + s[1:]
```
Comment (zhoxing-ms) [inline src/next/azext_next/utils.py:247]: Generally, please do not use a single character as the variable name, which will make the code difficult to understand

Comment (zhoxing-ms) [inline src/next/azext_next/utils.py:247]:
```suggestion
def capitalize_first_character_case(string):
    if not string:
        return string
    string = string.strip()
    return string[0].upper() + string[1:]
```

File: src/next/azext_next/custom.py (hunk 4 above)
Comment (ReaNAiveD) [reply src/next/azext_next/custom.py]: > Could we drop the usage of `colorama` as `Scenario Guide` does?

Done

File: src/next/azext_next/custom.py [hunk 10]
```diff
@@ -2,60 +2,97 @@
...
+                      (Style.SECONDARY, "(for example, enter \"b2\" for the second scenario. if none, enter 0)"),
+                      (Style.PRIMARY, ": ")]
+        group, option = select_combined_option(
+            option_msg,
+            {'a': OptionRange(1, len(command_recommendations)), 'b': OptionRange(1, len(scenario_recommendations))},
+            (None, -1))
+        if group == 'a':
+            rec = command_recommendations[option - 1]
+        elif group == 'b':
+            rec = scenario_recommendations[option - 1]
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:89]: We also need to send feedback data in these two cases

File: src/next/azext_next/custom.py [hunk 11]
```diff
@@ -307,18 +364,23 @@ def _execute_recommend_commands(cmd, rec):
 
 
 def _execute_recommend_scenarios(cmd, rec):
-    for nx_cmd in rec["nextCommandSet"]:
+    exec_idx = rec.get("executeIndex")
+    for idx, nx_cmd in enumerate(rec["nextCommandSet"]):
+        cmd_active = idx in exec_idx
+        if not cmd_active:
+            continue
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:371]: We may make the code simpler by traversing exec_idx first
```suggestion
    exec_idx = rec.get("executeIndex")
    for idx in exec_idx:
         nx_cmd = rec["nextCommandSet"][idx]
```

File: src/next/azext_next/custom.py [hunk 12]
```diff
@@ -150,26 +185,28 @@ def _give_recommend_commands(cmd, idx, rec):
...
 
 
 def _give_recommend_scenarios(idx, rec):
     index_str = "[" + str(idx) + "] "
-    num_notice = " ({} Commands)".format(len(rec['nextCommandSet']))
+    num_notice = f" ({len(rec['nextCommandSet'])} Commands)"
     print_styled_text([(Style.ACTION, index_str), (Style.PRIMARY, rec['scenario']), (Style.SECONDARY, num_notice)])
     if 'reason' in rec:
-        reason = rec['reason']
+        reason = rec['reason'].split('. ')[0]
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:202]:
```suggestion
        reason = rec['reason'].split('.')[0]
```

File: src/next/azext_next/custom.py (hunk 10 above)
Comment (ReaNAiveD) [reply src/next/azext_next/custom.py]: Done

File: src/next/azext_next/utils.py [hunk 13]
```diff
@@ -2,24 +2,50 @@
...
+        return default_value
+    if ret.isnumeric():
+        return None, int(ret)
+    while True:
+        group = ret[:1]
+        option = ret[1:]
+        if group not in group_range:
+            ret = input("The option should start with " + " or ".join(group_range) + ":")
+        elif not option.isnumeric():
+            ret = input("The option should end with a ledge number: ")
```
Comment (zhoxing-ms) [inline src/next/azext_next/utils.py:42]:
```suggestion
            ret = input("The option should end with a valid number: ")
```

File: src/next/azext_next/utils.py [hunk 14]
```diff
@@ -2,24 +2,50 @@
...
         if ret == '' or ret is None:
             return default_value
     return int(ret)
 
 
-def get_yes_or_no_option(option_description):
-    print(Fore.LIGHTBLUE_EX + ' ? ' + Fore.RESET + option_description, end='')
+def input_combined_option(group_range, default_value=(None, 0)):
+    """
+    Read combined option from stdin, ensure the group_name is in range or be None and option is a ledge number
```
Comment (zhoxing-ms) [inline src/next/azext_next/utils.py:25]:
```suggestion
    Read combined option from stdin, ensure the group_name is in range or be None and option is a valid number
```

File: src/next/azext_next/custom.py [hunk 15]
```diff
@@ -307,18 +367,21 @@ def _execute_recommend_commands(cmd, rec):
...
 
-        print("\nRunning: " + _get_command_item_sample(nx_cmd))
-        step_msg = "How do you want to run this step? 1. Run it 2. Skip it 3. Quit process " + Fore.LIGHTBLACK_EX \
-                   + "(Enter is to Run)" + Fore.RESET + ": "
-        run_option = get_int_option(step_msg, 1, 3, 1)
+        print_styled_text([(Style.ACTION, "Running: ")], end='')
+        print_styled_text(_get_command_sample(nx_cmd))
+        option_msg = [(Style.PRIMARY, "How do you want to run this step? 1. Run it 2. Skip it 3. Quit process "),
+                      (Style.SECONDARY, "(Enter is to Run)"), (Style.PRIMARY, ": ")]
+        run_option = select_option(option_msg, 1, 3, 1)
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:384]:
```suggestion
        run_option = select_option(option_msg, min_option=1, max_option=3, default_option=1)
```

File: src/next/azext_next/custom.py [hunk 16]
```diff
@@ -328,9 +391,10 @@ def _execute_recommend_scenarios(cmd, rec):
                     _print_help_info(cmd, nx_cmd["command"])
                     is_help_printed = True
 
-                step_msg = "Do you want to retry this step? 1. Run it 2. Skip it 3. Quit process " + Fore.LIGHTBLACK_EX \
-                           + "(Enter is to Run)" + Fore.RESET + ": "
-                run_option = get_int_option(step_msg, 1, 3, 1)
+                option_msg = [
+                    (Style.PRIMARY, "How do you want to run this step? 1. Run it 2. Skip it 3. Quit process "),
+                    (Style.SECONDARY, "(Enter is to Run)"), (Style.PRIMARY, ": ")]
+                run_option = select_option(option_msg, 1, 3, 1)
```
Comment (zhoxing-ms) [inline src/next/azext_next/custom.py:397]:
```suggestion
                run_option = select_option(option_msg, min_option=1, max_option=3, default_option=1)
```

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**